include/dsn/dist/distributed_lock_service.h

```diff
@@ -26,19 +26,38 @@
 # pragma once
 
 # include <dsn/service_api_cpp.h>
+# include <string>
+# include <functional>
 
 namespace dsn
 {
     namespace dist
     {
-        // interface
         class distributed_lock_service
         {
         public:
-            // void create_lock(..callback.);
-            // void lock(..callback.);
-            // void unlock(..callback.);
-            // void destroy_lock(..callback.);
+            typedef void* lock_handle;
+
+            virtual lock_handle create_lock(const std::string& lock_name) = 0;
+            /*
+             * try to get the lock:
+             *     lock_name: the global lock_id
+             *     myself: the name of myself. If someone got the lock
+             *       service_supplier use it to notify others who got the
+             *       lock
+             *     lock_callback:
+                     (1) if try_lock error, the lock_callback is called, and
+                         user can get the error reason by ec
+                     (2) if the caller don't get the lock, the lock_callback is called.
+                         In this case, ec should be 0, and caller could know who got the
+                         lock by the value of "who"
+                     (3) if the caller get the lock, the lock_callback is called
+                         and who==myself
+             */
+            virtual void try_lock(lock_handle handle,
+                                  const std::string& myself,
+                                  const std::function<void (int ec, std::string&& who)>& lock_callback) = 0;
+            virtual void free_lock(lock_handle handle) = 0;
```

Comment:
The conception is mixed in this case. A lock is shared and used by multiple clients - so I think in most cases, unlock is the right operation and destroy_lock is rare. In my opinion, there are two choices for this interface: (1).  open_lock(string, create_if_not_exist, cb), lock(h, cb), unlock(h, cb), close(h, destroy_or_not); or simpler version (2). lock(string, cb, create_if_not_exist); unlock(h, destroy_or_not).  Another issue is about leasing the locks when somebody is holding it. This is kinds of internal implementation of the provider, but still unlock may be done implicitly when the leasing operation is not done successfully due to maybe network failure. So cb for lock should be called: (1) when lock returns; (2) when lock lease failed. 

On the other hand, a callback should be present for all remote operations, as the network may fail, and you have to retry in that case, or even suicide in certain cases (e.g., leasing failed).